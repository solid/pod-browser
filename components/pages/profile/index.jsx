/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useContext } from "react";
import { SessionContext } from "@inrupt/solid-ui-react";
import { useRouter } from "next/router";
import Profile from "../../profile";
import EditableProfile from "../../profile/editableProfile";
import useFullProfile from "../../../src/hooks/useFullProfile";

export default function ProfileShow() {
  const router = useRouter();
  const { session, sessionRequestInProgress } = useContext(SessionContext);

  const decodedIri = router.query.webId
    ? decodeURIComponent(router.query.webId)
    : null;

  const agentProfile = useFullProfile(decodedIri ?? session.info.webId);

  if (sessionRequestInProgress || !agentProfile) {
    return null;
  }
  const editableProfile = agentProfile?.editableProfileDatasets[0];

  if (editableProfile) {
    return (
      <EditableProfile
        profile={agentProfile}
        profileDataset={editableProfile}
        webId={agentProfile.webId}
      />
    );
  }
  return <Profile profile={agentProfile} webId={decodedIri} />;
}
