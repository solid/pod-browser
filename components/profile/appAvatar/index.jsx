/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import { foaf } from "rdf-namespaces";
import { Avatar, Box, createStyles } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import { Icons } from "@inrupt/prism-react-components";
import {
  Text,
  ThingProvider,
  DatasetProvider,
  Image,
} from "@inrupt/solid-ui-react";

import styles from "./styles";
import {
  mockApp,
  mockAppDataset,
  LOGO_PREDICATE,
} from "../../../__testUtils/mockApp";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_NAME_TITLE = "app-name-title";
export const TESTCAFE_ID_NAME_FIELD = "app-name-field";
export const TESTCAFE_ID_WEBID_FIELD = "app-webid-field";
export const TESTCAFE_ID_TOS_FIELD = "app-tos-field";
export const TESTCAFE_ID_POLICY_FIELD = "app-policy-field";

export function setupErrorComponent(bem) {
  return () => (
    <Avatar className={bem("appAvatar")} alt="Contact avatar">
      <Icons className={bem("appAvatar")} name="project-diagram" />
    </Avatar>
  );
}

// temporarily using mock data for apps for dev purposes until we have audit list
const app = mockApp();
const dataset = mockAppDataset();

export default function AppAvatar() {
  const [error, setError] = useState(null);

  const classes = useStyles();
  const bem = useBem(classes);
  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  const errorComponent = setupErrorComponent(bem);

  return (
    <DatasetProvider solidDataset={dataset}>
      <ThingProvider thing={app} onError={setError}>
        <Box alignItems="center" display="flex">
          <Box>
            <Avatar className={classes.appAvatar} alt="Contact avatar">
              <Image
                property={LOGO_PREDICATE}
                width={120}
                errorComponent={errorComponent}
              />
            </Avatar>
          </Box>

          <Box p={2}>
            <h3 data-testid={TESTCAFE_ID_NAME_TITLE}>
              <Text property={foaf.name} />
            </h3>
          </Box>
        </Box>
      </ThingProvider>
    </DatasetProvider>
  );
}
