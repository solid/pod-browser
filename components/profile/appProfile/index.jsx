/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import { foaf, vcard } from "rdf-namespaces";
import clsx from "clsx";
import { getThing, getUrl } from "@inrupt/solid-client";
import {
  Avatar,
  Box,
  Paper,
  InputLabel,
  createStyles,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import {
  Container,
  Icons,
  Table as PrismTable,
} from "@inrupt/prism-react-components";
import {
  Text,
  ThingProvider,
  DatasetProvider,
  Value,
  Image,
  Table,
  TableColumn,
} from "@inrupt/solid-ui-react";

import { vcardExtras } from "../../../src/addressBook";

import styles from "./styles";
import {
  mockApp,
  mockAppDataset,
  TOS_PREDICATE,
  POLICY_PREDICATE,
  CONTACTS_PREDICATE,
  LOGO_PREDICATE,
} from "../../../__testUtils/mockApp";
import ValueBody from "./valueCellBody";
import TypeBody from "./titleCellBody";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_NAME_TITLE = "app-name-title";
export const TESTCAFE_ID_NAME_FIELD = "app-name-field";
export const TESTCAFE_ID_WEBID_FIELD = "app-webid-field";
export const TESTCAFE_ID_TOS_FIELD = "app-tos-field";
export const TESTCAFE_ID_POLICY_FIELD = "app-policy-field";

export function setupErrorComponent(bem) {
  return () => (
    <Avatar className={bem("appAvatar")} alt="Contact avatar">
      <Icons className={bem("appAvatar")} name="project-diagram" />
    </Avatar>
  );
}

// temporarily using mock data for apps for dev purposes until we have audit list
const app = mockApp();
const dataset = mockAppDataset();

export default function AppProfile() {
  const [error, setError] = useState(null);
  const tableClass = PrismTable.useTableClass("table", "inherits");

  const classes = useStyles();
  const bem = useBem(classes);
  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  const errorComponent = setupErrorComponent(bem);
  // FIXME: not sure what the contacts would look like in a real app, but possibly a dataset with things? Maybe a profile? For now it is an array of strings:
  const contactsUrl = getUrl(app, CONTACTS_PREDICATE);
  const contacts = getThing(dataset, contactsUrl);

  return (
    <Container>
      <Paper style={{ marginTop: "1em" }}>
        <Box p={2}>
          <DatasetProvider solidDataset={dataset}>
            <ThingProvider thing={app} onError={setError}>
              <Box alignItems="center" display="flex">
                <Box>
                  <Avatar className={classes.appAvatar} alt="Contact avatar">
                    <Image
                      property={LOGO_PREDICATE}
                      width={120}
                      errorComponent={errorComponent}
                    />
                  </Avatar>
                </Box>

                <Box p={2}>
                  <h3 data-testid={TESTCAFE_ID_NAME_TITLE}>
                    <Text property={foaf.name} />
                  </h3>
                </Box>
              </Box>

              <hr />

              <Box mt={2}>
                <Box>
                  <InputLabel>App WebID</InputLabel>
                  <Value
                    property={vcardExtras("WebId")}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_WEBID_FIELD}
                  />
                </Box>
              </Box>
              <Box mt={2}>
                <Box>
                  <InputLabel>Terms of Service</InputLabel>
                  <Value
                    property={TOS_PREDICATE}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_TOS_FIELD}
                  />
                </Box>
              </Box>
              <Box mt={2}>
                <Box>
                  <InputLabel>Privacy Policy</InputLabel>
                  <Value
                    property={POLICY_PREDICATE}
                    dataType="url"
                    autosave
                    data-testid={TESTCAFE_ID_POLICY_FIELD}
                  />
                </Box>
              </Box>
            </ThingProvider>
            <ThingProvider thing={contacts} onError={setError}>
              <Box mt={2}>
                <InputLabel>Contact Support</InputLabel>
                <Table
                  className={clsx(tableClass, bem("table"))}
                  things={[
                    { dataset, thing: contacts },
                    { dataset, thing: contacts },
                    { dataset, thing: contacts },
                  ]}
                >
                  <TableColumn
                    property={vcard.url}
                    body={({ row }) => <TypeBody id={row.id} />}
                    header=""
                  />
                  <TableColumn
                    dataType="url"
                    property={vcard.url}
                    header=""
                    body={({ row }) => <ValueBody id={row.id} />}
                  />
                </Table>
              </Box>
            </ThingProvider>
          </DatasetProvider>
        </Box>
      </Paper>
    </Container>
  );
}
