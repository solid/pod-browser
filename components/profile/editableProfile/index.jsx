/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import T from "prop-types";
import { Box, Paper } from "@material-ui/core";
import { Container } from "@inrupt/prism-react-components";
import { CombinedDataProvider } from "@inrupt/solid-ui-react";
import { getThing } from "@inrupt/solid-client";
import { schema } from "rdf-namespaces";
import PersonProfile from "../personProfile";
import PersonAvatar from "../personAvatar";
import AppProfile from "../appProfile";
import { profilePropTypes } from "../../../constants/propTypes";

export const TESTCAFE_ID_NAME_TITLE = "profile-name-title";
export const TESTCAFE_ID_NAME_FIELD = "profile-name-field";
export const TESTCAFE_ID_ROLE_FIELD = "profile-role-field";
export const TESTCAFE_ID_ORG_FIELD = "profile-org-field";

export default function EditableProfile({ profileDataset, profile, webId }) {
  if (!profileDataset) {
    return <span>No profile document found this this WebID</span>;
  }
  if (profile.types.includes(schema.SoftwareApplication)) {
    return <AppProfile />;
  }

  return (
    <CombinedDataProvider solidDataset={profileDataset} thingUrl={webId}>
      <Container>
        <Paper style={{ marginTop: "1em" }}>
          <Box p={2}>
            <PersonAvatar
              profileDataset={profileDataset}
              profile={profile}
              editing
            />
            <hr />
            <PersonProfile
              profileDataset={profileDataset}
              profile={profile}
              editing
            />
          </Box>
        </Paper>
      </Container>
    </CombinedDataProvider>
  );
}

EditableProfile.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  profileDataset: T.object,
  profile: profilePropTypes,
  webId: T.string.isRequired,
};

EditableProfile.defaultProps = {
  profileDataset: null,
  profile: null,
};
