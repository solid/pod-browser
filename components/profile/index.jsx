/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import T from "prop-types";
import { Box, Paper } from "@material-ui/core";
import { Alert } from "@material-ui/lab";
import { Container } from "@inrupt/prism-react-components";
import { CombinedDataProvider } from "@inrupt/solid-ui-react";
import { schema } from "rdf-namespaces";
import { isHTTPError } from "../../src/error";
import PersonProfile from "./personProfile";
import PersonAvatar from "./personAvatar";
import AppProfile from "./appProfile";

export const TESTCAFE_ID_NAME_TITLE = "profile-name-title";
export const TESTCAFE_ID_NAME_FIELD = "profile-name-field";
export const TESTCAFE_ID_ROLE_FIELD = "profile-role-field";
export const TESTCAFE_ID_ORG_FIELD = "profile-org-field";

export default function Profile(props) {
  const { profileIri, editing, type } = props;
  const [error, setError] = useState(null);

  // TODO replace with toast error or something?
  if (error) {
    if (isHTTPError(error, 404)) {
      return (
        <Alert severity="error">
          {`Cannot fetch avatar for this WebID: ${profileIri}`}
        </Alert>
      );
    }
    return error.toString();
  }

  if (type === schema.SoftwareApplication) {
    return <AppProfile />;
  }

  return (
    <CombinedDataProvider
      datasetUrl={profileIri}
      thingUrl={profileIri}
      onError={setError}
    >
      <Container>
        <Paper style={{ marginTop: "1em" }}>
          <Box p={2}>
            <PersonAvatar profileIri={profileIri} />
            <hr />
            <PersonProfile profileIri={profileIri} editing={editing} />
          </Box>
        </Paper>
      </Container>
    </CombinedDataProvider>
  );
}

Profile.propTypes = {
  profileIri: T.string.isRequired,
  editing: T.bool,
  type: T.string.isRequired,
};

Profile.defaultProps = {
  editing: false,
};
