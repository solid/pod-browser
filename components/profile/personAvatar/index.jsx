/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import T from "prop-types";
import { foaf, vcard } from "rdf-namespaces";
import { Avatar, Box, createStyles } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import { Text, Image, useSession } from "@inrupt/solid-ui-react";

import styles from "./styles";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_NAME_TITLE = "profile-name-title";

export function setupErrorComponent(bem) {
  return () => (
    <Avatar className={bem("avatar")} alt="Contact photo placeholder" />
  );
}

export default function PersonAvatar({ profileIri }) {
  const classes = useStyles();
  const bem = useBem(classes);
  const errorComponent = setupErrorComponent(bem);
  const { session } = useSession();

  return (
    <Box alignItems="center" display="flex">
      <Box>
        <Avatar className={classes.avatar}>
          <Image
            property={vcard.hasPhoto}
            width={120}
            alt={profileIri}
            errorComponent={errorComponent}
          />
        </Avatar>
      </Box>

      <Box p={2}>
        <h3 data-testid={TESTCAFE_ID_NAME_TITLE}>
          <Text className={classes.avatarText} property={foaf.name} />
          <a
            className={classes.headerLink}
            href={profileIri}
            rel="noreferrer"
            target="_blank"
          >
            {session.info.webId}
          </a>
        </h3>
      </Box>
    </Box>
  );
}

PersonAvatar.propTypes = {
  profileIri: T.string.isRequired,
};
