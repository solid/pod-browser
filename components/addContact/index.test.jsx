/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { foaf } from "rdf-namespaces";
import {
  addStringNoLocale,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import mockSession from "../../__testUtils/mockSession";
import mockSessionContextProvider from "../../__testUtils/mockSessionContextProvider";
import AddContact, {
  EXISTING_WEBID_ERROR_MESSAGE,
  FETCH_PROFILE_FAILED_ERROR_MESSAGE,
  handleSubmit,
} from "./index";
import * as addressBookFns from "../../src/addressBook";
import { contactsContainerIri } from "../../src/addressBook";
import { renderWithTheme } from "../../__testUtils/withTheme";
import {
  aliceWebIdUrl,
  mockPersonDatasetAlice,
  mockProfileAlice,
} from "../../__testUtils/mockPersonResource";
import * as profileHelperFns from "../../src/solidClientHelpers/profile";

describe("AddContact", () => {
  test("it renders the Add Contact form", () => {
    const session = mockSession();
    const SessionProvider = mockSessionContextProvider(session);
    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <AddContact />
      </SessionProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
describe("handleSubmit", () => {
  const addressBookUri = "https://example.com/addressBook";
  const addressBook = mockSolidDatasetFrom(addressBookUri);
  let setAgentId;
  let setIsLoading;
  let alertError;
  let alertSuccess;
  let setDirtyForm;
  let session;

  beforeEach(() => {
    setAgentId = jest.fn();
    setIsLoading = jest.fn();
    alertError = jest.fn();
    alertSuccess = jest.fn();
    setDirtyForm = jest.fn();
    session = mockSession();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test("it alerts the user and exits if the webid already exists", async () => {
    const personDataset = mockPersonDatasetAlice();
    const personProfile = mockProfileAlice();
    const people = [personDataset];
    const handler = handleSubmit({
      addressBook,
      setAgentId,
      setIsLoading,
      alertError,
      alertSuccess,
      session,
      setDirtyForm,
      people,
    });
    jest
      .spyOn(profileHelperFns, "fetchProfile")
      .mockResolvedValue(personProfile);
    jest
      .spyOn(addressBookFns, "findContactInAddressBook")
      .mockResolvedValue([personDataset]);
    await handler("alreadyExistingWebId");
    expect(addressBookFns.findContactInAddressBook).toHaveBeenCalledWith(
      people,
      aliceWebIdUrl,
      session.fetch
    );
    expect(setAgentId).toHaveBeenCalledTimes(0);
    expect(setIsLoading).toHaveBeenCalledTimes(2);
    expect(alertError).toHaveBeenCalledWith(EXISTING_WEBID_ERROR_MESSAGE);
    expect(setDirtyForm).toHaveBeenCalledWith(true);
  });

  test("it alerts the user and exits if fetching the profile fails", async () => {
    const mockProfileError = new Error("error");
    const handler = handleSubmit({
      addressBook,
      setAgentId,
      setIsLoading,
      alertError,
      alertSuccess,
      session,
      setDirtyForm,
    });
    jest
      .spyOn(profileHelperFns, "fetchProfile")
      .mockRejectedValueOnce(mockProfileError);
    await handler("iri");
    expect(setAgentId).toHaveBeenCalledTimes(0);
    expect(setIsLoading).toHaveBeenCalledTimes(2);
    expect(alertSuccess).not.toHaveBeenCalled();
    expect(alertError).toHaveBeenCalledWith(FETCH_PROFILE_FAILED_ERROR_MESSAGE);
  });

  test("it saves a contact without name and alerts the user", async () => {
    const personUri = "http://example.com/marie#me";
    const mockProfile = { webId: personUri };
    const handler = handleSubmit({
      addressBook,
      setAgentId,
      setIsLoading,
      alertError,
      alertSuccess,
      session,
      setDirtyForm,
    });
    jest.spyOn(profileHelperFns, "fetchProfile").mockResolvedValue(mockProfile);
    jest
      .spyOn(addressBookFns, "findContactInAddressBook")
      .mockResolvedValue(undefined);
    jest.spyOn(addressBookFns, "saveContact").mockResolvedValue({
      response: "peopleDataset",
      error: null,
    });
    await handler(personUri);
    expect(setAgentId).toHaveBeenCalledTimes(1);
    expect(setIsLoading).toHaveBeenCalledTimes(2);
    expect(alertSuccess).toHaveBeenCalled();
  });

  test("it saves a new contact", async () => {
    const personUri = "http://example.com/alice#me";
    const personDataset = addStringNoLocale(
      mockThingFrom(personUri),
      foaf.name,
      "Alice"
    );
    const contactsIri = contactsContainerIri("http://www.example.com/");
    const people = mockSolidDatasetFrom(contactsIri);
    const peopleMutate = jest.fn();
    const peopleDatasetWithContact = setThing(people, personDataset);
    const mockProfile = mockProfileAlice();
    const handler = handleSubmit({
      addressBook,
      setAgentId,
      setIsLoading,
      alertError,
      alertSuccess,
      session,
      setDirtyForm,
      people,
      peopleMutate,
    });
    jest
      .spyOn(addressBookFns, "findContactInAddressBook")
      .mockResolvedValue(undefined);
    jest.spyOn(profileHelperFns, "fetchProfile").mockResolvedValue(mockProfile);
    jest.spyOn(addressBookFns, "saveContact").mockResolvedValue({
      response: peopleDatasetWithContact,
      error: null,
    });
    await handler(personUri);
    expect(setAgentId).toHaveBeenCalledWith("");
    expect(setIsLoading).toHaveBeenCalledTimes(2);
    expect(alertSuccess).toHaveBeenCalledWith(
      "Alice was added to your contacts"
    );
    expect(alertError).not.toHaveBeenCalled();
    expect(setDirtyForm).toHaveBeenCalledWith(false);
  });

  test("it alerts the user if there is an error while creating the contact", async () => {
    const mockProfile = mockProfileAlice();
    const handler = handleSubmit({
      addressBook,
      setAgentId,
      setIsLoading,
      alertError,
      alertSuccess,
      session,
      setDirtyForm,
    });
    jest.spyOn(profileHelperFns, "fetchProfile").mockResolvedValue(mockProfile);
    jest
      .spyOn(addressBookFns, "findContactInAddressBook")
      .mockResolvedValue(undefined);
    jest.spyOn(addressBookFns, "saveContact").mockResolvedValue({
      response: null,
      error: "There was an error saving the resource",
    });
    await handler(aliceWebIdUrl);
    expect(setAgentId).toHaveBeenCalledTimes(0);
    expect(setIsLoading).toHaveBeenCalledTimes(2);
    expect(alertSuccess).not.toHaveBeenCalled();
    expect(alertError).toHaveBeenCalledWith(
      "There was an error saving the resource"
    );
  });
});
