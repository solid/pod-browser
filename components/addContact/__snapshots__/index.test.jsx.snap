// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddContact renders the Add Contact form 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-container"
  >
    <nav
      class="PodBrowser-back-to-nav"
    >
      <span
        class="PodBrowser-icon-caret-left PodBrowser-back-to-nav__caret"
      />
      <a
        class="PodBrowser-back-to-nav__link"
        href="/contacts"
      >
        Back to contacts
      </a>
    </nav>
    <h3>
      Add new contact
    </h3>
    <form
      class="PodBrowser-form"
    >
      <label
        class="PodBrowser-label"
      >
        WebID
      </label>
      <input
        class="PodBrowser-input"
        id="id1"
        pattern="https:\\\\/\\\\/\\\\S+"
        title="Must be a valid URL that starts with https:// and does not contain spaces"
        type="url"
        value=""
      />
      <button
        class="PodBrowser-button"
        data-testid="add-agent-button"
        type="submit"
      >
        <span
          class="PodBrowser-button__text"
        >
          Add Contact
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
